website/source/api/features/filtering.html.md

```diff
@@ -0,0 +1,458 @@
+base 10 integers and floating point numbers are possible. For quoted strings,
+they may either be enclosed in double quotes or backticks. When enclosed in
+backticks they are treated as raw strings and escape sequences such as `\n`
+will not be expanded.
+
+## Filter Utilization
+
+Generally, only the main object is filtered. When filtering for
+an item within an array, that is not at the top level, the entire array that contains the item
+will be returned. This is usually the outermost object of a response,
+but in some cases such the [`/catalog/node/:node`](api/catalog.html#list-services-for-node)
+endpoint the filtering is performed on a object embedded within the results.
+
+### Performance
+
+Filters are executed on the servers and therefore will consume some amount
+of CPU time on the server. For non-stale queries this means that the filter
+is executed on the leader.
+
+### Filtering Examples
+
+#### Agent API
+
+**Command - Unfiltered**
+
+```sh
+curl -X GET localhost:8500/v1/agent/services
+```
+
+**Response - Unfiltered**
+
+```json
+{
+    "redis1": {
+        "ID": "redis1",
+        "Service": "redis",
+        "Tags": [
+            "primary",
+            "production"
+        ],
+        "Meta": {
+            "env": "production",
+            "foo": "bar"
+        },
+        "Port": 1234,
+        "Address": "",
+        "Weights": {
+            "Passing": 1,
+            "Warning": 1
+        },
+        "EnableTagOverride": false
+    },
+    "redis2": {
+        "ID": "redis2",
+        "Service": "redis",
+        "Tags": [
+            "secondary",
+            "production"
+        ],
+        "Meta": {
+            "env": "production",
+            "foo": "bar"
+        },
+        "Port": 1235,
+        "Address": "",
+        "Weights": {
+            "Passing": 1,
+            "Warning": 1
+        },
+        "EnableTagOverride": false
+    },
+    "redis3": {
+        "ID": "redis3",
+        "Service": "redis",
+        "Tags": [
+            "primary",
+            "qa"
+        ],
+        "Meta": {
+            "env": "qa"
+        },
+        "Port": 1234,
+        "Address": "",
+        "Weights": {
+            "Passing": 1,
+            "Warning": 1
+        },
+        "EnableTagOverride": false
+    }
+}
+```
+
+**Command - Filtered**
+
+```sh
+curl -G localhost:8500/v1/agent/services --data-urlencode 'filter=Meta.env == qa'
+```
+
+**Response - Filtered**
+
+```json
+{
+    "redis3": {
+        "ID": "redis3",
+        "Service": "redis",
+        "Tags": [
+            "primary",
+            "qa"
+        ],
+        "Meta": {
+            "env": "qa"
+        },
+        "Port": 1234,
+        "Address": "",
+        "Weights": {
+            "Passing": 1,
+            "Warning": 1
+        },
+        "EnableTagOverride": false
+    }
+}
+```
+
+#### Catalog API
+
+**Command - Unfiltered**
+
+```sh
+curl -X GET localhost:8500/v1/catalog/service/api-internal
```

Comment:
Also maybe we shouldn't give an example of using an internal endpoint.

Reply:
This isn't an internal endpoint. I just had some mock services in a consul instance and one was called api-internal. I could change this if you still think its confusing.